File: .claude/skills/agent-skill-kit-commons/packages/py/src/skills_ref/prompt/budget.py
from __future__ import annotations

_HEADER_FOOTER_CHARS = len("<available_skills>\n</available_skills>\n")
_PER_ENTRY_OVERHEAD = len('  <skill name="" tier=""></skill>\n')


def truncate_to_budget(entries, max_tokens):
    if not max_tokens:
        return list(entries), 0
    max_chars = max_tokens * 4
    used = _HEADER_FOOTER_CHARS
    kept = []
    for e in entries:
        cost = (
            _PER_ENTRY_OVERHEAD
            + len(e.get("name") or "")
            + len(e.get("tier") or "")
            + len(e.get("description") or "")
        )
        if used + cost > max_chars:
            break
        used += cost
        kept.append(e)
    return kept, len(entries) - len(kept)

File: .claude/skills/agent-skill-kit-commons/packages/mjs/src/prompt/budget.mjs
const HEADER_FOOTER_CHARS = "<available_skills>\n</available_skills>\n".length;
const PER_ENTRY_OVERHEAD = '  <skill name="" tier=""></skill>\n'.length;

export function truncateToBudget(entries, maxTokens) {
  if (!maxTokens) return { kept: entries, dropped: 0 };
  const maxChars = maxTokens * 4;
  let used = HEADER_FOOTER_CHARS;
  const kept = [];
  for (const e of entries) {
    const cost =
      PER_ENTRY_OVERHEAD +
      (e.name?.length ?? 0) +
      (e.tier?.length ?? 0) +
      (e.description?.length ?? 0);
    if (used + cost > maxChars) break;
    used += cost;
    kept.push(e);
  }
  return { kept, dropped: entries.length - kept.length };
}
